vignettes/datatable-joins.Rmd

```diff
@@ -0,0 +1,687 @@
+
+```{r}
+Products = data.table(
+  id = c(1:4, 
+         NA_integer_),
+  name = c("banana",
+           "carrots",
+           "popcorn",
+           "soda",
+           "toothpaste"),
+  price = c(0.63,
+            0.89,
+            2.99,
+            1.49,
+            2.99),
+  unit = c("unit",
+           "lb",
+           "unit",
+           "ounce",
+           "unit"),
+  type = c(rep("natural", 2L),
+           rep("processed", 3L))
+)
+
+Products
+```
+
+2. Defining a `data.table` showing the proportion of taxes to be applied for processed products based on their units.
+
+```{r}
+NewTax = data.table(
+  unit = c("unit","ounce"),
+  type = "processed",
+  tax_prop = c(0.65, 0.20)
+)
+
+NewTax
+```
+
+
+3. Defining a `data.table` simulating the products received every Monday with a `product_id` that is not present in the `Products` table.
+
+```{r}
+set.seed(2156)
+
+ProductReceived = data.table(
+  id = 1:10,
+  date = seq(from = as.IDate("2024-01-08"), length.out = 10L, by = "week"), 
+  product_id = sample(c(NA_integer_, 1:3, 6L), size = 10L, replace = TRUE),
+  count = sample(c(50L, 100L, 150L), size = 10L, replace = TRUE)
+)
+
+ProductReceived
+```
+
+4. Defining a `data.table` to show some sales that can take place on weekdays with another `product_id` that is not present in the `Products` table.
+
+```{r}
+sample_date = function(from, to, size, ...){
+  all_days = seq(from = from, to = to, by = "day")
+  weekdays = all_days[wday(all_days) %in% 2:6]
+  days_sample = sample(weekdays, size, ...)
+  days_sample_desc = sort(days_sample)
+  return(days_sample_desc)
+}
+
+set.seed(5415)
+
+ProductSales = data.table(
+  id = 1:10,
+  date = ProductReceived[, sample_date(min(date), max(date), 10L)], 
+  product_id = sample(c(1:3, 7L), size = 10L, replace = TRUE),
+  count = sample(c(50L, 100L, 150L), size = 10L, replace = TRUE)
+)
+
+
+ProductSales
+```
+
+## 2. `data.table` joining syntax 
+
+Before taking advantage of the `data.table` syntax to perform join operations we need to know which arguments can help us to perform successful joins.
+
+The next diagram shows a description for each basic argument. In the following sections we will show how to use each of them and add more complexity little by little.
+
+```
+x[i, on, nomatch]
+| |  |   |
+| |  |   ----> If NULL only returns rows linked in x and i tables
+| |  ----> a character vector o list defining match logict
+| ----> principal data.table, list or data.frame
+----> secundary data.table
+```
+
+An important difference between the regular `data.table` syntax is that _the only argument you can pass by position is the `i` argument_, the rest as you will see are going to be passed by name, so **feel free to change the argument order any time using the argument names** if it seems more convenient.
```

Comment:
this is not true. In R you can pass any argument by name or position. on is 15th so it is difficult, but possible.

Reply:
I understand your point.

Here is the new text

> Please keep in mind that the standard argument order in data.table is `dt[i, j, by]`. For join operations, it is recommended to pass the `on` and `nomatch` arguments by name to avoid using `j` and `by` when they are not needed.

Reply:
here is a demonstration that it is possible to use on argument by position instead of name (I don't think this is a great idea to do in practice though haha)
```r
> dt1=data.table(x=1:2, y=3:4)
> dt2=data.table(x=2)
> dt1[dt2,,,,,,,,,,,,,,"x"]
       x     y
   <int> <int>
1:     2     4
```